gapis/trace/android/trace.go

```diff
@@ -0,0 +1,487 @@
+	"github.com/google/gapid/core/os/android/apk"
+	"github.com/google/gapid/core/os/device/bind"
+	"github.com/google/gapid/gapidapk"
+	"github.com/google/gapid/gapidapk/pkginfo"
+	gapii "github.com/google/gapid/gapii/client"
+	"github.com/google/gapid/gapis/trace/tracer"
+)
+
+// Only update the package list every 30 seconds at most
+var packageUpdateTime = 30.0
+
+type androidTracer struct {
+	b                    adb.Device
+	packages             *pkginfo.PackageList
+	lastIconDensityScale float32
+	lastPackageUpdate    time.Time
+}
+
+func (t *androidTracer) GetPackages(ctx context.Context, isRoot bool, iconDensityScale float32) (*pkginfo.PackageList, error) {
+	refresh := time.Since(t.lastPackageUpdate).Seconds() > packageUpdateTime ||
+		t.lastIconDensityScale != iconDensityScale
+
+	if t.packages != nil && !isRoot {
+		refresh = false
+	}
+	if refresh {
+		packages, err := gapidapk.PackageList(ctx, t.b, true, iconDensityScale)
+		if err != nil {
+			return nil, err
+		}
+		pkgList := &pkginfo.PackageList{
+			Packages:       []*pkginfo.Package{},
+			Icons:          packages.Icons,
+			OnlyDebuggable: packages.OnlyDebuggable,
+		}
+
+		for _, p := range packages.Packages {
+			for _, activity := range p.Activities {
+				if len(activity.Actions) > 0 {
+					pkgList.Packages = append(pkgList.Packages, p)
+					break
+				}
+			}
+		}
+
+		t.packages = pkgList
+		t.lastPackageUpdate = time.Now()
+		t.lastIconDensityScale = iconDensityScale
+	}
+	return t.packages, nil
+}
+
+// NewTracer returns a new Tracer for Android.
+func NewTracer(dev bind.Device) tracer.Tracer {
+	return &androidTracer{dev.(adb.Device), nil, 1.0, time.Time{}}
+}
+
+// IsServerLocal returns true if all paths on this device can be server-local
+func (t *androidTracer) IsServerLocal() bool {
+	return false
+}
+
+func (t *androidTracer) CanSpecifyCWD() bool {
+	return false
+}
+
+func (t *androidTracer) CanSpecifyEnv() bool {
+	return false
+}
+
+func (t *androidTracer) CanUploadApplication() bool {
+	return true
+}
+
+func (t *androidTracer) HasCache() bool {
+	return true
+}
+
+func (t *androidTracer) CanUsePortFile() bool {
+	return false
+}
+
+func (t *androidTracer) PreferredRootUri(ctx context.Context) (string, error) {
+	return "", nil
+}
+
+func (t *androidTracer) APITraceOptions(ctx context.Context) []tracer.APITraceOptions {
+	options := make([]tracer.APITraceOptions, 0, 2)
+	if t.b.Instance().Configuration.Drivers.Opengl.Version != "" {
+		options = append(options, tracer.GLESTraceOptions())
+	}
+	if len(t.b.Instance().Configuration.Drivers.Vulkan.PhysicalDevices) > 0 {
+		options = append(options, tracer.VulkanTraceOptions())
+	}
+	return options
+}
+
+func (t *androidTracer) GetTraceTargetNode(ctx context.Context, uri string, iconDensity float32) (*tracer.TraceTargetTreeNode, error) {
+	packages, err := t.GetPackages(ctx, uri == "", iconDensity)
+
+	if err != nil {
+		return nil, err
+	}
+	if uri == "" {
+		r := &tracer.TraceTargetTreeNode{
+			"",
```

Comment:
Please use named fields when building a struct from an external package. This is a govet or golint warning.